### PR TITLE
Fix oncefrom with MachineConfig

### PR DESCRIPTION
**Pull request by runcom**
**- What I did**

Builds on top of #464 so we can be sure the node calls are at a minimum required. Looks like onceFrom with a MachineConfig still requires a cluster to be around (not the case for raw ignition).
The patch makes sure a kube and mc clients are around when running in onceFrom with a machineconfig avoiding any panic.

This also incorporates https://github.com/openshift/machine-config-operator/pull/476 and it's now correctly passing.

/cc @ashcrow ptal

**- How to verify it**

**- Description for the changelog**

### CODE REVIEW

**Comment by runcom** [comment]
/hold 

holding cause it requires #464

**Comment by runcom** [comment]
The daemons are failing with something related to https://github.com/openshift/machine-config-operator/pull/463:
```
E0223 19:11:07.625913    5505 daemon.go:444] Fatal error checking initial state of node: Checking initial state: parsing booted osImageURL: invalid reference format
E0223 19:11:07.625940    5505 writer.go:90] Marking degraded due to: Checking initial state: parsing booted osImageURL: invalid reference format
I0223 19:11:07.636545    5505 daemon.go:446] Entering degraded state; going to sleep
```

**Comment by runcom** [comment]
:thinking: 

/retest

**Comment by runcom** [comment]
/hold cancel

rebased and ready to be reviewed

**Comment by runcom** [comment]
origin flakes

/retest

**Comment by runcom** [comment]
openshift-sdn flake (?) - reported

/retest

File: cmd/machine-config-daemon/start.go [hunk 1]
```diff
@@ -81,13 +85,39 @@ func runStartCmd(cmd *cobra.Command, args []string) {
...
+		ctx *common.ControllerContext
+	)
 	glog.Info("starting node writer")
 	nodeWriter := daemon.NewNodeWriter()
 	go nodeWriter.Run(stopCh)
 
+	cb, err := common.NewClientBuilder(startOpts.kubeconfig)
+	if err != nil {
+		if startOpts.onceFrom != "" {
+			glog.Info("Cannot initialize ClientBuilder, likely in onceFrom mode with Ignition")
```
Comment (cgwalters) [inline cmd/machine-config-daemon/start.go]: "likely"?  We just checked it :wink: 

Why not just

```
if startopts.onceFrom == "" {
   glog.Fatalf(...)
}
```

File: cmd/machine-config-daemon/start.go [hunk 2]
```diff
@@ -81,13 +85,39 @@ func runStartCmd(cmd *cobra.Command, args []string) {
...
+
+	var (
+		kubeClient kubernetes.Interface
+		mcClient   mcfgclientset.Interface
+	)
+
+	if cb != nil {
+		kubeClient, err = cb.KubeClient(componentName)
+		if err != nil {
+			glog.Info("Cannot initialize kubeClient, likely in onceFrom mode with Ignition")
```
Comment (cgwalters) [inline cmd/machine-config-daemon/start.go]: Confused by this, won't `cb == nil` if we're in oncefrom?

File: pkg/daemon/daemon.go [hunk 3]
```diff
@@ -149,6 +143,14 @@ const (
 	updateDelay = 5 * time.Second
 )
 
+type onceFromOrigin int
+
+const (
+	onceFromUnknownConfig onceFromOrigin = iota
+	onceFromLocalConfig
+	onceFromRemoteConfig
```
Comment (cgwalters) [inline pkg/daemon/daemon.go]: I like this!

**Comment by cgwalters** [comment]
I rebased this :surfing_man:

**Comment by cgwalters** [comment]
@vrutkovs One thing that would help us...are there use cases for *both* oncefrom with raw Ignition and from a MachineConfig object pulled from the cluster?  Why do we need both?

**Comment by vrutkovs** [comment]
>are there use cases for both oncefrom with raw Ignition and from a MachineConfig object pulled from the cluster? Why do we need both?

In scaleup we only use Ignition, can't think of a case where MachineConfig + --once-from would be used

**Comment by sdodson** [comment]
/retest
VPC limits

**Comment by cgwalters** [comment]
We can address the nits above in followup.

/lgtm

File: cmd/machine-config-daemon/start.go (hunk 1 above)
Comment (runcom) [reply cmd/machine-config-daemon/start.go]: we haven't checked that we're in onceFrom _with_ ignition, that's why, the log is confusing though, need to reword it

**Pull request merged**